tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -305,142 +305,110 @@ public:
    }
 
    void InitTask(TTreeReader *, unsigned int) {}
-
-   void Exec(unsigned int slot, double x0) // 1D histos
-   {
-      fObjects[slot]->Fill(x0);
+   
+   // all scalar arguments
+   template <typename... ValTypes,
+            typename std::enable_if<Conjunction<std::is_convertible<ValTypes, double>...>::value, int>::type = 0>
+   void Exec(unsigned int slot, ValTypes... x) {
+      fObjects[slot]->Fill(x...);
    }
 
-   void Exec(unsigned int slot, double x0, double x1) // 1D weighted and 2D histos
-   {
-      fObjects[slot]->Fill(x0, x1);
-   }
+   // helper functions which provide one implementation for scalar types and another for containers
+   // TODO these could probably all be replaced by inlined lambdas and/or constexpr if statements
+   // in c++17 or later
 
-   void Exec(unsigned int slot, double x0, double x1, double x2) // 2D weighted and 3D histos
-   {
-      fObjects[slot]->Fill(x0, x1, x2);
-   }
+   // return unchanged value for scalar
+   template<typename T, typename std::enable_if<std::is_convertible<T, double>::value, int>::type = 0>
```

Comment:
This is because in some cases the check is for scalar vs container, and for other cases the check is for scalar vs iterator, so in this way the same test can be used everywhere.